src/scheduler/job_info.c

```diff
@@ -1896,6 +1893,467 @@ translate_fail_code(schd_error *err, char *comment_msg, char *log_msg)
+ * @param policy[in] - policy info
+ * @param sinfo[in] - server universe
+ * @return resdef **
+ * @retval array of resdefs for creating resresv_set's resources.
+ * @retval NULL on error
+ */
+resdef **
+create_resresv_sets_resdef(status *policy, server_info *sinfo) {
+	resdef **defs;
+	int ct;
+	int i;
+	if (policy == NULL || sinfo == NULL)
+		return NULL;
+	
+	ct = count_array((void **) policy->resdef_to_check);
+	/* 6 for ctime, walltime, max_walltime, min_walltime, preempt_targets (maybe), and NULL*/
+	defs = malloc((ct+6) * sizeof(resdef *));
+	
+	for (i = 0; i < ct; i++)
+		defs[i] = policy->resdef_to_check[i];
+	defs[i++] = getallres(RES_CPUT);
+	defs[i++] = getallres(RES_WALLTIME);
+	defs[i++] = getallres(RES_MAX_WALLTIME);
+	defs[i++] = getallres(RES_MIN_WALLTIME);
+	if(sinfo->preempt_targets_enable)
+		defs[i++] = getallres(RES_PREEMPT_TARGETS);
+	defs[i] = NULL;
+	
+	return defs;
+}
+
+/**
+ * @brief create a resresv_set based on a resource_resv
+ *
+ * @param[in] policy - policy info
+ * @param[in] sinfo - server info
+ * @param[in] resresv - resresv to create resresv set from
+ *
+ * @return resresv_set **
+ * @retval newly created resresv_set
+ * @retval NULL on error
+ */
+resresv_set *
+create_resresv_set_by_resresv(status *policy, server_info *sinfo, resource_resv *resresv)
+{
+	resresv_set *rset;
+	if (policy == NULL || resresv == NULL)
+		return NULL;
+
+	rset = new_resresv_set();
+	if (rset == NULL)
+		return NULL;
+
+	if (resresv_set_use_user(sinfo))
+		rset->user = string_dup(resresv->user);
+	if (resresv_set_use_grp(sinfo))
+		rset->group = string_dup(resresv->group);
+	if (resresv_set_use_proj(sinfo))
+		rset->project = string_dup(resresv->project);
+	rset->select_spec = dup_selspec(resresv->select);
+	rset->place_spec = dup_place(resresv->place_spec);
+	rset->req = dup_selective_resource_req_list(resresv->resreq, policy->equiv_class_resdef);
+	if (resresv->is_job && resresv->job != NULL) {
+		if (resresv_set_use_queue(resresv->job->queue))
+			rset->qinfo = resresv->job->queue;
+	}
+
+	return rset;
+}
+
+/**
+ * @brief find the index of a resresv_set by its component parts
+ * @par each parameter can be NULL if the resresv_set does not have one
+ * @param[in] rsets - resresv_sets to search
+ * @param[in] user - user name
+ * @param[in] group - group name
+ * @param[in] project - project name
+ * @param[in] sel - select spec
+ * @param[in] req - list of resources (i.e., qsub -l)
+ * @param[in] qinfo - queue
+ * @return int
+ * @retval index of resresv if found
+ * @retval -1 if not found or on error
+ */
+int
+find_resresv_set(status *policy, resresv_set **rsets, char *user, char *group, char *project, selspec *sel, place *pl, resource_req *req, queue_info *qinfo)
+{
+	int i;
+
+	if (rsets == NULL)
+		return -1;
+
+	for (i = 0; rsets[i] != NULL; i++) {
+		if (qinfo != NULL && rsets[i]->qinfo == NULL || qinfo == NULL && rsets[i]->qinfo != NULL)
```

Comment:
There are two checks for queue.  The first is just a quick if one is NULL and the other is not.  If that's the case then the resresv_set doesn't match.  Once the first check has passed, either both are NULL or both are not NULL.  The second check will either pass through if both are NULL or do a strcmp() on the queue names if they aren't.